src/re/multi_grid/indexing.py

```diff
@@ -0,0 +1,969 @@
+        splits = tuple(np.atleast_1d(s) for s in splits)
+        return self.__class__(
+            shape0=self.shape0, splits=self.splits + splits, atLevel=self.atLevel
+        )
+
+    def at(self, level: int):
+        level = self._parse_level(level)
+        if level > 0:
+            fct = np.array(
+                [reduce(operator.mul, si) for si in zip(*self.splits[:level])]
+            )
+        else:
+            fct = 1
+        return self.atLevel(
+            shape=self.shape0 * fct,
+            splits=self.splits[level] if level < self.depth else None,
+            parent_splits=self.splits[level - 1] if level >= 1 else None,
+        )
+
+
+@dataclass()
+class OpenGridAtLevel(GridAtLevel):
+    padding: Optional[npt.NDArray[np.int_]]
+    parent_padding: Optional[npt.NDArray[np.int_]]
+    shifts: Optional[npt.NDArray[np.int_]]
+
+    def __init__(
+        self,
+        shape,
+        splits=None,
+        parent_splits=None,
+        padding=None,
+        parent_padding=None,
+        shifts=None,
+    ):
+        if padding is not None:
+            self.padding = np.atleast_1d(padding)
+        if parent_padding is not None:
+            self.parent_padding = np.atleast_1d(parent_padding)
+        if shifts is not None:
+            self.shifts = np.atleast_1d(shifts)
+        super().__init__(shape=shape, splits=splits, parent_splits=parent_splits)
+
+    def refined_indices(self):
+        if self.splits is None:
+            raise IndexError("This level has no children")
+        # TODO non-dense grid
+        return np.mgrid[
+            tuple(slice(pp, sh - pp) for sh, pp in zip(self.shape, self.padding))
+        ]
+
+    def children(self, index):
+        if (self.splits is None) or (self.padding is None):
+            raise IndexError("This level has no children")
+        lo = self.padding[(slice(None),) + (np.newaxis,) * (index.ndim - 1)]
+        hi = self.shape[(slice(None),) + (np.newaxis,) * (index.ndim - 1)] - lo
+        # TODO add option for fully jax transformable error handling
+        # Follow jax-array inspired out of bounds handling
+        index = select(index >= lo, index, lo * jnp.ones_like(index))
+        index = select(index < hi, index, (hi - 1) * jnp.ones_like(index))
+        return super().children(index - lo)
+
+    def neighborhood(self, index, window_size: Iterable[int]):
+        index = self._parse_index(index)
+        dtp = np.result_type(index)
+        window_size = np.asarray(window_size)
+        assert window_size.size == self.ndim
+        c = np.mgrid[tuple(slice(sz) for sz in window_size)]
+        c -= (window_size // 2)[(slice(None),) + (np.newaxis,) * self.ndim]
+        c_bc = (
+            (slice(None),)
+            + (np.newaxis,) * (index.ndim - 1)
+            + (slice(None),) * self.ndim
+        )
+        id_bc = (slice(None),) * index.ndim + (np.newaxis,) * self.ndim
+        res = index[id_bc] + c[c_bc]
```

Comment:
The only difference that I could spot compared to the `GridAtLevel` implementation is the missing modulo. Is it really that weird if we would still wrap indices around? If we choose to still wrap indices around we could use `super().neighborhood`... :)